librocksdb_sys/crocksdb/c.cc

```diff
@@ -2915,4 +2923,270 @@ const char* crocksdb_pinnableslice_value(const crocksdb_pinnableslice_t* v,
+                      entry_type, seq, file_size);
+    return Status::OK();
+  }
+
+  virtual Status Finish(UserCollectedProperties* inner) override {
+    crocksdb_user_collected_properties_t props(inner);
+    rep_->finish(rep_, &props);
+    return Status::OK();
+  }
+
+  virtual UserCollectedProperties GetReadableProperties() const override {
+    UserCollectedProperties inner;
+    crocksdb_user_collected_properties_t props(&inner);
+    rep_->readable_properties(rep_, &props);
+    return inner;
+  }
+
+  const char* Name() const override {
+    return rep_->name(rep_);
+  }
+};
+
+struct crocksdb_table_properties_collector_factory_context_t {
+  void* factory;
+  const char* (*name)(void*);
+  void (*destructor)(void*);
+  crocksdb_table_properties_collector_context_t*
+  (*create_table_properties_collector)(void*, uint32_t cf);
+};
+
+struct crocksdb_table_properties_collector_factory_t : public TablePropertiesCollectorFactory {
+  crocksdb_table_properties_collector_factory_context_t* rep_;
+
+  crocksdb_table_properties_collector_factory_t(
+    crocksdb_table_properties_collector_factory_context_t* context) : rep_(context) {}
+
+  virtual ~crocksdb_table_properties_collector_factory_t() {
+    rep_->destructor(rep_);
+  }
+
+  virtual TablePropertiesCollector* CreateTablePropertiesCollector (
+    TablePropertiesCollectorFactory::Context ctx) override {
+    auto context = rep_->create_table_properties_collector(rep_, ctx.column_family_id);
+    return new crocksdb_table_properties_collector_t(context);
+  }
+
+  const char* Name() const override {
+    return rep_->name(rep_);
+  }
+};
+
+crocksdb_table_properties_collector_factory_t*
+crocksdb_table_properties_collector_factory_create(
+  crocksdb_table_properties_collector_factory_context_t* context) {
+  return new crocksdb_table_properties_collector_factory_t(context);
+}
+
+void crocksdb_table_properties_collector_factory_destroy(
+  crocksdb_table_properties_collector_factory_t* factory) {
+  delete factory;
+}
+
+void crocksdb_options_add_table_properties_collector_factory(
+  crocksdb_options_t* opt, crocksdb_table_properties_collector_factory_t* factory) {
+  opt->rep.table_properties_collector_factories.push_back(
+    std::shared_ptr<TablePropertiesCollectorFactory>(factory));
+}
+
+struct crocksdb_shallow_strings_map_t {
+  size_t size = 0;
+  const char** keys = nullptr;
+  size_t* keys_lens = nullptr;
+  const char** values = nullptr;
+  size_t* values_lens = nullptr;
+
+  ~crocksdb_shallow_strings_map_t() {
+    delete[] keys;
+    delete[] keys_lens;
+    delete[] values;
+    delete[] values_lens;
+  }
+
+  void assign(const std::map<std::string, std::string>& items) {
+    size = items.size();
+    keys = new const char* [size];
+    keys_lens = new size_t [size];
+    values = new const char* [size];
+    values_lens = new size_t [size];
+    size_t i = 0;
+    for (auto it = items.begin(); it != items.end(); it++, i++) {
+      keys[i] = it->first.data();
```

Comment:
The life time of `items` must be longer than `keys/values`, is this can be guaranteed?